Code/Engine/GameEngine/Gameplay/Implementation/RaycastComponent.cpp

```diff
@@ -0,0 +1,190 @@
+
+ezRaycastComponent::~ezRaycastComponent() = default;
+
+
+void ezRaycastComponent::OnSimulationStarted()
+{
+  m_pPhysicsWorldModule = GetWorld()->GetModule<ezPhysicsWorldModuleInterface>();
+  m_hLastTriggerObjectInRay.Invalidate();
+}
+
+void ezRaycastComponent::SerializeComponent(ezWorldWriter& stream) const
+{
+  SUPER::SerializeComponent(stream);
+  auto& s = stream.GetStream();
+
+  stream.WriteGameObjectHandle(m_hRaycastEndObject);
+  s << m_fMaxDistance;
+  s << m_bDisableTargetObjectOnNoHit;
+  s << m_uiCollisionLayerEndPoint;
+  s << m_uiCollisionLayerTrigger;
+  s << m_sTriggerMessage;
+}
+
+void ezRaycastComponent::DeserializeComponent(ezWorldReader& stream)
+{
+  SUPER::DeserializeComponent(stream);
+  auto& s = stream.GetStream();
+
+  m_hRaycastEndObject = stream.ReadGameObjectHandle();
+  s >> m_fMaxDistance;
+  s >> m_bDisableTargetObjectOnNoHit;
+  s >> m_uiCollisionLayerEndPoint;
+  s >> m_uiCollisionLayerTrigger;
+  s >> m_sTriggerMessage;
+}
+
+
+void ezRaycastComponent::SetTriggerMessage(const char* sz)
+{
+  m_sTriggerMessage.Assign(sz);
+}
+
+const char* ezRaycastComponent::GetTriggerMessage() const
+{
+  return m_sTriggerMessage.GetData();
+}
+
+void ezRaycastComponent::SetRaycastEndObject(const char* szReference)
+{
+  auto resolver = GetWorld()->GetGameObjectReferenceResolver();
+
+  if (!resolver.IsValid())
+    return;
+
+  m_hRaycastEndObject = resolver(szReference, GetHandle(), "RaycastEndObject");
+}
+
+void ezRaycastComponent::Update()
+{
+  if (m_hRaycastEndObject.IsInvalidated())
+    return;
+
+  ezGameObject* pEndObject = nullptr;
+  if (!GetWorld()->TryGetObject(m_hRaycastEndObject, pEndObject))
+  {
+    return;
+  }
+
+  if (!m_pPhysicsWorldModule)
+  {
+    // Happens in Prefab viewports
+    return;
+  }
+
+  const ezVec3 rayStartPosition = GetOwner()->GetGlobalPosition();
+  const ezVec3 rayDir = GetOwner()->GetGlobalDirForwards();
+
+  //ezDebugRenderer::Line lines[] = { {rayStartPosition, rayStartPosition + rayDir * m_fMaxDistance} };
+  //ezDebugRenderer::DrawLines(GetWorld(), lines, ezColor::RebeccaPurple);
+
+  float fHitDistance = m_fMaxDistance;
+  ezPhysicsHitResult hit;
+  if (m_pPhysicsWorldModule->CastRay(rayStartPosition, rayDir, m_fMaxDistance, m_uiCollisionLayerEndPoint, hit))
+  {
+    fHitDistance = hit.m_fDistance;
+
+    if (!pEndObject->IsActive() && m_bDisableTargetObjectOnNoHit)
+    {
+      pEndObject->SetActive(true);
+    }
+  }
+  else
+  {
+    if (m_bDisableTargetObjectOnNoHit)
+    {
+      pEndObject->SetActive(false);
+    }
+    else
+    {
+      if (!pEndObject->IsActive())
+      {
+        pEndObject->SetActive(true);
+      }
+    }
+  }
+
+  if (!m_sTriggerMessage.IsEmpty() && m_uiCollisionLayerEndPoint != m_uiCollisionLayerTrigger)
+  {
+    ezPhysicsHitResult triggerHit;
+    if (m_pPhysicsWorldModule->CastRay(rayStartPosition, rayDir, m_fMaxDistance, m_uiCollisionLayerTrigger, triggerHit) && triggerHit.m_fDistance < fHitDistance)
```

Comment:
Instead of m_fMaxDistance you could use the distance of the previous hit, you are only interested in things that are closer.

Reply:
If I have the situation (A ray start, B wall hit):
A --- 1 --- 2 --- B
And I move 1 out of the ray the distance increases but I still want to trigger that I hit 2 again.

Reply:
I assumed you make the raycast for static stuff first, so you will hit B first. Then you do the second raycast, now you only need the distance to B, because you can only hit 1 or 2, if they are closer than B.

Reply:
ah, right - that should've been fHitDistance, not m_fMaxDistance.